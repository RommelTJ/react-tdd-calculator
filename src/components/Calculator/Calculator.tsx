import React, { Component } from 'react';

import './Calculator.css';
import Display from '../Display/Display';
import Keypad from '../Keypad/Keypad';

class Calculator extends Component {
  state = {
    // value to be displayed in <Display />
    displayValue: '0',
    // values to be displayed in number <Keys />
    numbers: ['9', '8', '7', '6', '5', '4', '3', '2', '1', '.', '0','ce'],
    // values to be displayed in operator <Keys />
    operators: ['/', 'x', '-', '+'],
    // operator selected for math operation
    selectedOperator: '',
    // stored value to use for math operation
    storedValue: '',
  }

  callOperator = () => {
    console.log('call operation');
    let { displayValue, selectedOperator, storedValue } = this.state;
    // temp variable for updating state storedValue
    const updateStoredValue = displayValue;

    // parse strings for operations
    let displayValueNumber = parseInt(displayValue, 10);
    const storedValueNumber = parseInt(storedValue, 10);

    // performs selected operation
    switch (selectedOperator) {
      case '+':
        displayValueNumber = storedValueNumber + displayValueNumber;
        break;
      case '-':
        displayValueNumber = storedValueNumber - displayValueNumber;
        break;
      case 'x':
        displayValueNumber = storedValueNumber * displayValueNumber;
        break;
      case '/':
        displayValueNumber = storedValueNumber / displayValueNumber;
        break;
      default:
        // set displayValue to zero if no case matches
        displayValueNumber = 0;
    }

    // converts displayValue to a string
    displayValue = displayValueNumber.toString();
    // reset selectedOperator
    selectedOperator = '';
    // check for 'NaN' or 'Infinity', if true set displayValue to '0'
    if (displayValue === 'NaN' || displayValue === 'Infinity') displayValue = '0';

    this.setState({ displayValue, selectedOperator, storedValue: updateStoredValue });
  }

  setOperator = (value: string) => {
    console.log('set operation');
    let { displayValue, selectedOperator, storedValue } = this.state;

    // check if a value is already present for selectedOperator
    if (selectedOperator === '') {
      // update storedValue to the value of displayValue
      storedValue = displayValue;
      // reset the value of displayValue to '0'
      displayValue = '0';
      // update the value of selectedOperator to the given value
      selectedOperator = value;
    } else {
      // if selectedOperator is not an empty string
      // update the value of selectedOperator to the given value
      selectedOperator = value;
    }

    this.setState({ displayValue, selectedOperator, storedValue });
  }

  updateDisplay = (value: string) => {
    console.log('update display');
    let { displayValue } = this.state;

    // prevent multiple occurrences of '.'
    if (value === '.' && displayValue.includes('.')) value = '';

    if (value === 'ce') {
      // deletes last char in displayValue
      displayValue = displayValue.substr(0, displayValue.length - 1);
      // set displayValue to '0' if displayValue is empty string
      if (displayValue === '') displayValue = '0';
    } else {
      // replace displayValue with value if displayValue equal to '0'
      // else concatenate displayValue and value
      displayValue === '0' ? displayValue = value : displayValue += value;
    }

    this.setState({ displayValue });
  }

  render = () => {
    const { displayValue, numbers, operators } = this.state;

    return (
      <div data-testid="calculator" role="presentation" className="calculator-container">
        <Display displayValue={displayValue} />
        <Keypad
          callOperator={this.callOperator}
          numbers={numbers}
          operators={operators}
          setOperator={this.setOperator}
          updateDisplay={this.updateDisplay}
        />
      </div>
    );
  }
}

export default Calculator;
