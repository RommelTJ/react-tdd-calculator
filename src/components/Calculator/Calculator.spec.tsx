import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react'

import Calculator from './Calculator';
import display from "../Display/Display";

describe('Calculator', () => {
  let wrapper: RenderResult;

  beforeEach(() => wrapper = render(<Calculator />));

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should render a div', () => {
    const divs = screen.getAllByRole('presentation');
    expect(divs).toHaveLength(1);
  });

  it('should render the Display and Keypad Components', () => {
    const displayComponent = screen.getByTestId('display');
    expect(displayComponent).toBeTruthy();
    const keypadComponent = screen.getByTestId('keypad');
    expect(keypadComponent).toBeTruthy();
  });

});
