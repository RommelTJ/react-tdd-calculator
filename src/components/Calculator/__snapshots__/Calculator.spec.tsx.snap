// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator should render correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="calculator-container"
        data-testid="calculator"
        role="presentation"
      >
        <div
          class="display-container"
          data-testid="display"
        >
          <p
            class="display-value"
          >
            0
          </p>
        </div>
        <div
          class="keypad-container"
          data-testid="keypad"
        >
          <div
            class="numbers-container"
          />
          <div
            class="operators-container"
          />
          <div
            class="submit-container"
          >
            <div
              class="key-container submit-key"
              data-testid="key-="
            >
              <p
                class="key-value"
              >
                =
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="calculator-container"
      data-testid="calculator"
      role="presentation"
    >
      <div
        class="display-container"
        data-testid="display"
      >
        <p
          class="display-value"
        >
          0
        </p>
      </div>
      <div
        class="keypad-container"
        data-testid="keypad"
      >
        <div
          class="numbers-container"
        />
        <div
          class="operators-container"
        />
        <div
          class="submit-container"
        >
          <div
            class="key-container submit-key"
            data-testid="key-="
          >
            <p
              class="key-value"
            >
              =
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
